system/ehci.c

```diff
@@ -337,12 +335,12 @@ static void build_ehci_qtd(ehci_qtd_t *this_qtd, const ehci_qtd_t *final_qtd, ui
     this_qtd->control       = EHCI_QTD_CPAGE(0) | EHCI_QTD_CERR(3) | control;
     this_qtd->data_length   = dt | (length & 0x7fff);
     this_qtd->buffer_ptr[0] = (uintptr_t)buffer;
+
+    memset((char *)this_qtd + offsetof(ehci_qtd_t, buffer_ptr[1]), 0, sizeof(ehci_qtd_t) - offsetof(ehci_qtd_t, buffer_ptr[1]));
```

Comment:
I don't understand that change in ehci.c, but if it's OK for @martinwhitaker, that's fine.

Reply:
The previous fields were already initialized one at a time and in the right order, after the call to `memset()` located at the beginning of the function. This `memset()` was therefore effectively doing superfluous work.

Reply:
Yes, but the original code is much easier to read and to see is correct.
Does this actually reduce the code size?

Reply:
Sure, in fact ISTR that the out of line full memset -> reduced __builtin_memset change in that function even makes for the largest individual size optimization in that `memset()`-related commit :)

Reply:
I just tested in a 64-bit build, and the changes in ehci.c (separate from the changes in string.{c,h} reduce the program size by 16 bytes, although most of that is likely due to alignment - disassembly shows that `build_ehci_qtd` is 2 bytes smaller and `build_ehci_qhd` is the same size. I'd rather have the more readable code.

Reply:
Here on a 64-bit build, with GCC 11 (Debian sid amd64):
* before the first commit containing the memset -> __builtin_memset change: `0000000000002488 l     F .text  0000000000000070 build_ehci_qtd`
* after the first commit, i.e memset as. __builtin_memset + memset(x + offsetof, 0, size - offsetof) at the end of the function: `0000000000002482 l     F .text  000000000000004a build_ehci_qtd`
* memset as. __builtin_memset + memset(x, 0, size) at the beginning of the function: `0000000000002482 l     F .text  000000000000004f build_ehci_qtd`
I'll take the third version nevertheless.